README.md
# NameSilo Let's Encrypt

[![License (3-Clause BSD)](https://img.shields.io/badge/license-BSD%203--Clause-blue.svg?style=flat-square)](http://opensource.org/licenses/BSD-3-Clause)
[![Python 3.4](https://img.shields.io/badge/python-3.4-blue.svg)](https://www.python.org/)

Python scripts (hook) to automate obtaining [Let's Encrypt](https://letsencrypt.org/) certificates,
using [Certbot](https://certbot.eff.org/) DNS-01 challenge validation for domains DNS hosted on
[NameSilo](https://www.namesilo.com/).

## Setup

The scripts use the [tldextract](https://github.com/john-kurkowski/tldextract) and [untangle](https://untangle.readthedocs.io/en/latest/) libraries, if not already installed on your system:

```
pip install tldextract untangle
```

Download the [latest release](https://github.com/ethauvin/namesilo-letsencrypt/releases) archive and expand it in the desired directory.


## Configuration

Add your [NameSilo API key](https://www.namesilo.com/account_api.php)
to the top of the `config.py` file:

```python
# Get your API Key from: https://www.namesilo.com/account_api.php
apikey = "YOUR_API_KEY"
```

Alternatively, the API key can be set in the `NAMESILO_API` environment variable.

## Using with Certbot

To issue or renew a certificate using the hook scripts, try something like:

<pre>
certbot certonly --manual --email you@example.com \
--agree-tos --manual-public-ip-logging-ok \
--preferred-challenges=dns \
--manual-auth-hook <em>/path/to/authenticator.py</em> \
--manual-cleanup-hook <em>/path/to/cleanup.py</em> \
-d *.example.com -d example.com
</pre>

Please note that NameSilo DNS propagation takes up to **15 minutes**. The scripts will wait **25 minutes** before completing, just to be safe.


CHANGELOG.md
# Changelog

## [0.9.2](https://github.com/ethauvin/namesilo-letsencrypt/tree/0.9.2) (2020-05-28)

[Full Changelog](https://github.com/ethauvin/namesilo-letsencrypt/compare/0.9.1...0.9.2)

**Closed issues:**

- Do not support multiple \_acme-challenge records [\#1](https://github.com/ethauvin/namesilo-letsencrypt/issues/1)

**Merged pull requests:**

- make python3.4 compatible. add subdomains support [\#2](https://github.com/ethauvin/namesilo-letsencrypt/pull/2) ([Junker](https://github.com/Junker))

## [0.9.1](https://github.com/ethauvin/namesilo-letsencrypt/tree/0.9.1) (2020-03-15)

[Full Changelog](https://github.com/ethauvin/namesilo-letsencrypt/compare/0.9.0...0.9.1)

## [0.9.0](https://github.com/ethauvin/namesilo-letsencrypt/tree/0.9.0) (2019-04-30)

[Full Changelog](https://github.com/ethauvin/namesilo-letsencrypt/compare/8a91146faebd196b223f7653e470428130f26246...0.9.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


cleanup.py
#!/usr/bin/env python3

#  cleanup.py
#
#  Copyright (c) 2019-2020, Erik C. Thauvin (erik@thauvin.net)
#  All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#
#    Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
#    Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
#    Neither the name of this project nor the names of its contributors may be
#    used to endorse or promote products derived from this software without
#    specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
#  THE POSSIBILITY OF SUCH DAMAGE.

import os
import sys
import tempfile
import urllib.request

import tldextract
import untangle

from config import apikey

domain = os.environ['CERTBOT_DOMAIN']
tmpdir = os.path.join(tempfile.gettempdir(), "CERTBOT_" + domain)

if "NAMESILO_API" in os.environ:
    apikey = os.environ['NAMESILO_API']

tld = tldextract.extract(domain)
nsdomain = tld.domain + "." + tld.suffix

url = "https://www.namesilo.com/api/dnsDeleteRecord\
?version=1&type=xml&key=" + apikey + "&domain=" + nsdomain + "&rrid="


def getrequest(record_id):
    return urllib.request.Request(
        url + record_id,
        data=None,
        headers={
            'User-Agent': ('Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:74.0) '
                           'Gecko/20100101 Firefox/74.0')
        }
    )


idFile = os.path.join(tmpdir, "RECORD_ID")
if os.path.isfile(idFile):
    f = open(idFile, "r")
    for line in f:
        with urllib.request.urlopen(getrequest(line.rstrip())) as response:
            html = response.read()
        xml = untangle.parse(str(html, 'utf-8'))
        if xml.namesilo.reply.code.cdata != '300':
            print("{}: {} ({})".format(
                domain,
                xml.namesilo.reply.detail.cdata,
                xml.namesilo.reply.code.cdata), file=sys.stderr)
            if xml.namesilo.reply.code.cdata != '280':
                sys.exit(1)
    f.close()
    os.remove(idFile)


authenticator.py
#!/usr/bin/env python3

#  authenticator.py
#
#  Copyright (c) 2019-2020, Erik C. Thauvin (erik@thauvin.net)
#  All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#
#    Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
#    Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
#    Neither the name of this project nor the names of its contributors may be
#    used to endorse or promote products derived from this software without
#    specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
#  THE POSSIBILITY OF SUCH DAMAGE.

import os
import sys
import tempfile
import time
import urllib.request

import tldextract
import untangle

from config import apikey, wait


def sleep(minutes):
    if minutes < 16:
        minutes = 16
    time.sleep(minutes * 60)


domain = os.environ['CERTBOT_DOMAIN']
validation = os.environ['CERTBOT_VALIDATION']
tmpdir = os.path.join(tempfile.gettempdir(), "CERTBOT_" + domain)
rrhost = "_acme-challenge"

if "NAMESILO_API" in os.environ:
    apikey = os.environ['NAMESILO_API']

tld = tldextract.extract(domain)
nsdomain = tld.domain + "." + tld.suffix
if tld.subdomain:
    rrhost += "." + tld.subdomain

url = "https://www.namesilo.com/api/dnsAddRecord?\
version=1&type=xml&key=" + apikey + "&domain=" + nsdomain + "&rrtype=TXT\
&rrhost=" + rrhost + "&rrvalue=" + validation + "&rrttl=3600"

req = urllib.request.Request(
    url,
    data=None,
    headers={
        'User-Agent': ('Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:74.0) '
                       'Gecko/20100101 Firefox/74.0')
    }
)

with urllib.request.urlopen(req) as response:
    html = response.read()
xml = untangle.parse(str(html, 'utf-8'))

if not os.path.exists(tmpdir):
    os.mkdir(tmpdir, 0o700)

if xml.namesilo.reply.code.cdata == '300':
    f = open(os.path.join(tmpdir, "RECORD_ID"), "a+")
    print(xml.namesilo.reply.record_id.cdata, file=f)
    f.close()
else:
    print("{}: {} ({})".format(domain,
                               xml.namesilo.reply.detail.cdata,
                               xml.namesilo.reply.code.cdata), file=sys.stderr)
    sys.exit(1)

# Sleep X minutes
sleep(wait)
